### PR TITLE
Fix ExaAdmmBackend for CPU and GPU

### PR DESCRIPTION
**Pull request by AnirudhSubramanyam**
* Fixes the interface to ExaAdmm
* Supports an additional type of contingency linking constraint with ExaAdmm ("preventive" SCOPF)
* Adds tests for ExaAdmm on GPU
* Minor updates to MPI tests
* Minor updates to ProxAL constructors

### CODE REVIEW

File: src/Evaluators/NonDecomposedModel.jl [hunk 1]
```diff
@@ -28,6 +28,28 @@ function NonDecomposedModel(
...
+Instantiate non-decomposed multi-period ACOPF instance
+using `rawdata` with model parameters
+`modelinfo` and algorithm parameters `algparams`.
+
+"""
+function NonDecomposedModel(
+    rawdata::RawData,
+    modelinfo::ModelInfo,
+    algparams::AlgParams,
+    space::AbstractBackend=JuMPBackend(),
```
Comment (frapac) [inline src/Evaluators/NonDecomposedModel.jl:51]: Can we solve the `NonDecomposedModel` with another backend than `JuMPBackend`? If not, I would remove this option and force using JuMP here. And specify this explicitly in the documentation.

File: src/ExaAdmmBackend/ExaAdmmBackend.jl [hunk 2]
```diff
@@ -39,25 +39,12 @@ mutable struct ModelProxAL{T,TD,TI,TM} <: ExaAdmm.AbstractOPFModel{T,TD,TI,TM}
     nvar_padded::Int
 
     # ProxAL's part
-    t_curr::Int      # current time period
```
Comment (frapac) [inline src/ExaAdmmBackend/ExaAdmmBackend.jl:42]: Good to know we can get rid of all this info. The simplest, the better.

However, where are we accounting for the different time step (origin, middle and final)? Directly when setting the coefficients for `Q_ref` and `c_ref`?

File: src/ExaAdmmBackend/interface.jl [hunk 3]
```diff
@@ -22,24 +22,22 @@ macro define_setter_value(function_name, attribute)
...
 end
+function set_generator_cost!(model::ModelProxAL)
+    ngen = model.grid_data.ngen
+    baseMVA = model.grid_data.baseMVA
+    Q_ref = zero(model.Q_ref)
+    c_ref = zero(model.c_ref)
+    Q_ref[1:4:4*ngen] .= 2.0 * model.grid_data.c2[:] * baseMVA^2
+    c_ref[1:2:2*ngen] .= model.grid_data.c1[:] * baseMVA
+    copyto!(model.Q_ref, Q_ref)
+    copyto!(model.c_ref, c_ref)
```
Comment (frapac) [inline src/ExaAdmmBackend/interface.jl:39]: That's indeed better

File: src/ExaAdmmBackend/interface.jl [hunk 4]
```diff
@@ -22,24 +22,22 @@ macro define_setter_value(function_name, attribute)
     end
 end
 
-@define_setter_array set_lower_bound_slack! smin
```
Comment (frapac) [inline src/ExaAdmmBackend/interface.jl:25]: I am not sure to understand where we set `smin` and `smax` now?

File: src/backends.jl [hunk 5]
```diff
@@ -608,45 +606,43 @@ function AdmmBlockBackend(
...
+    ExaAdmmBackend.set_active_load!(opfmodel, opfdata.Pd[:, 1])
+    ExaAdmmBackend.set_reactive_load!(opfmodel, opfdata.Qd[:, 1])
 
     # Set bounds on slack variables s
-    if algparams.decompCtgs && modelinfo.ctgs_link_constr_type == :corrective_penalty && k > 1
-        ExaAdmmBackend.set_upper_bound_slack!(opfmodel, 2 .* [g.scen_agc for g in gens])
+    if algparams.decompCtgs && k > 1
+        if modelinfo.ctgs_link_constr_type == :corrective_penalty
+            copyto!(opfmodel.smin, zeros(length(gens)))
+            copyto!(opfmodel.smax, 2.0.*[g.scen_agc for g in gens])
```
Comment (frapac) [inline src/backends.jl:637]: Ok, nice!

**Review by frapac — APPROVED** [review]
This PR is a net improvement, and remove dead code for good. Ok to merge!

**Comment by michel2323** [comment]
> Alternatively, we only test using 2 MPI ranks and then everything passes anyway since those 2 ranks will exchange everything between each other. What do you think?
> 

@AnirudhSubramanyam I was just finding out the same. We should either error out or deal with it. I'm gonna take a look tomorrow.

**Comment by michel2323** [comment]
I've just realized that these lines are a problem when running at scale with high T and K https://github.com/exanauts/ProxAL.jl/blob/c0a117b04aa9ae45068f56358c4318cf4db32340/src/OPF/opfsolution.jl#L30-L38


In fact, it's normal that only that process whose work it was has the solution for that block. @AnirudhSubramanyam I'd say for the PR above we fix it for now by setting np=2 and then think this over.

Opened an issue:
https://github.com/exanauts/ProxAL.jl/issues/68

File: src/Evaluators/NonDecomposedModel.jl (hunk 1 above)
Comment (AnirudhSubramanyam) [reply src/Evaluators/NonDecomposedModel.jl:51]: Good idea

File: src/ExaAdmmBackend/ExaAdmmBackend.jl (hunk 2 above)
Comment (AnirudhSubramanyam) [reply src/ExaAdmmBackend/ExaAdmmBackend.jl:42]: Yes, indeed!

**Pull request merged**